File: Python/Programmers/Level 2/피보나치 수.py
# 재귀함수는 depth 제한이 있으므로 DP 알고리즘 적용
# f(n) = f(n-1) + f(n-2), f(0) = f(1) = 1

# def solution(n):
#     dp = [0] * (n) # 피보나치 수를 저장할 DP 테이블
#     dp[0] = dp[1] = 1 # 첫 번째와 2번 째 값 저장
#     for i in range(2, n):
#         dp[i] = dp[i-1] + dp[i-2] # 3번째값부터 점화식 적용
        
#     return dp[n-1] % 1234567

def solution(n):
    a, b = 0, 1
    for i in range(n):
        a, b = b, a+b # 파이썬의 특성을 이용하여 피보나치 수 구하기

    return a % 1234567

File: JavaScript/Programmers/Level 2/피보나치 수.js
// 문제에서 주어진대로 n번째 피보나치 수를 구하면됨
// n의 범위가 크기 때문에 재귀로 풀면 시간 초과
// 다이나믹 프로그래밍 알고리즘 적용
function solution(n) {
    const dp = Array(n + 1).fill(0); // dp 테이블 생성
    [dp[0], dp[1]] = [0, 1]; // 초기값 설정
    for (let i = 2; i <= n; i++) {
        // 2번째 요소부터 점화식대로 적용
        // 이때 숫자가 계속해서 커지기 때문에 연산이 진행될 때마다 나머지 연산 진행
        dp[i] = (dp[i - 1] + dp[i - 2]) % 1234567;
    }
    
    return dp[n];
}
